crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -46,13 +49,72 @@ pub enum Event {
     /// The receiver should return true (if the contract exists) or false (if it does not exist)
     /// for each contract using the [oneshot::channel].
     QueryContractExistance(Vec<ClassHash>, oneshot::Sender<Vec<bool>>),
+    /// A new L2 pending update was polled.
+    Pending(Box<PendingBlock>, Box<sequencer::reply::StateUpdate>),
+}
+
+/// Poll's the Sequencer's pending block and emits [Event::Pending]
+/// until the pending block is no longer connected to our current head.
+///
+/// This disconnect is detected whenever
+/// - `pending.block_hash != head`, or
+/// - `pending` is a fully formed block and not [PendingBlock]
+async fn poll_pending(
+    tx_event: &mpsc::Sender<Event>,
+    sequencer: &impl sequencer::ClientApi,
+    head: StarknetBlockHash,
+    poll_interval: Duration,
+) -> anyhow::Result<()> {
+    use crate::core::BlockId;
+
+    loop {
+        use sequencer::reply::MaybePendingBlock;
+
+        let block = match sequencer
+            .block(BlockId::Pending)
+            .await
+            .context("Download block")?
+        {
+            MaybePendingBlock::Block(block) => {
+                tracing::debug!(hash=%block.block_hash, "Found full block, exiting pending mode.");
```

Comment:
Can't we reuse this L2 accepted block instead of re-downloading it when switching from "poll pending" to "get head" mode?

Reply:
Yes one could; but I really don't want to add the plumbing for this into the current spaghetti.

Reply:
I was thinking the same, seems wasteful but thought of the likely response :)

Reply:
Let's note it down as an issue? If there's a big refactor which solves this, we'll close the issue, otherwise we'll have a reminder to do it even later :wink:  .